uvtools/dspec.py

```diff
@@ -146,6 +147,250 @@ def delay_filter(data, wgts, bl_len, sdf, standoff=0., horizon=1., min_dly=0.0,
+        coefficients, or a slower generalized least-squares solver. 
+        Default: True.
+    
+    operator : array_like, optional
+        Fourier basis operator matrix. Must have shape (Nfreq, Nmodes), where 
+        Nmodes = 2*nmax + 1.
+    
+    Returns
+    -------
+    model : array_like
+        Best-fit model, composed of a sum of Fourier modes.
+        
+    model_coeffs : array_like
+        Coefficients of Fourier modes, ordered from modes [-n, +n].
+    
+    data_out : array_like
+        In-painted data.
+    """
+    # Construct Fourier basis operator if not specified
+    if operator is None:
+        # Frequency array
+        nu = np.arange(data.size)
+        L = nu[-1] - nu[0]
+        
+        # Build matrix operator for complex Fourier basis
+        n = np.arange(-nmax, nmax+1)
+        F = np.array([np.exp(-1.j * _n * nu / L) for _n in n])
+    else:
+        F = operator
+        n = np.arange(-nmax, nmax+1)
+        assert F.shape[0] == 2*nmax+1, "Fourier basis operator has the wrong shape"
+    
+    # Turn flags into a mask
+    w = np.logical_not(flags)
+    
+    # Define model and likelihood function
+    model = lambda cn: np.dot(cn, F)
+    
+    def loglike(cn):
+        # Need to do real and imaginary parts separately, otherwise leastsq() fails
+        _delta = w * (data - model(cn[:n.size] + 1.j*cn[n.size:]))
+        delta = np.concatenate((_delta.real/sigma, _delta.imag/sigma))
+        return -0.5 * delta**2.
+    
+    # Initial guess for Fourier coefficients (real + imaginary blocks)
+    cn_in = np.zeros(2*n.size)
+    if cn_guess is not None:
+        assert cn_in.size == 2*cn_guess.size, "cn_guess must be of size %s" \
+            % (cn_in.size/2)
+        cn_in[:cn_guess.shape[0]] = cn_guess.real
+        cn_in[cn_guess.shape[0]:] = cn_guess.imag
+    
+    # Run least-squares fit
+    if use_linear:
+        # Solve as linear system
+        A = np.atleast_2d(w).T * F.T
+        res = lsq_linear(A/sigma, w*data/sigma)
+        cn_out = res.x
+    else:
+        # Use full non-linear leastsq fit
+        cn, stat = leastsq(loglike, cn_in)
+        cn_out = cn[:n.size] + 1.j*cn[n.size:]
+    
+    # Inject smooth best-fit model into masked areas
+    bf_model = model(cn_out)
+    data_out = data.copy()
+    data_out[flags] = bf_model[flags]
+    
+    # Add noise to in-painted regions if requested
+    if add_noise:
+        noise = np.random.randn(np.sum(flags)) + 1.j * np.random.randn(np.sum(flags))
+        if isinstance(sigma, np.ndarray):
+            data_out[flags] += sigma[flags] * noise
+        else:
+            data_out[flags] += sigma * noise
+    
+    # Return coefficients and best-fit model
+    return bf_model, cn_out, data_out
+
+
+def delay_filter_leastsq(data, flags, sigma, nmax, add_noise=False, 
+                         cn_guess=None, use_linear=True):
+    """
+    Fit a smooth model to 2D complex-valued data with flags, using a linear 
```

Comment:
I think it would be useful to mention that under the hood this just calls `delay_filter_leastsq_1d` along the first axis of the `data` array (i.e., it's not doing a two-dimensional fit).

Reply:
Done!